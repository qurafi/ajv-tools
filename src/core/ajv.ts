import Ajv from "ajv";
import generateAjvStandaloneCode from "ajv/dist/standalone/index.js";
import { red, yellow } from "kleur/colors";
import { createDebug, removeSchemaFileExt } from "../utils/index.js";
import type { Options as AjvOptions } from "ajv";
import rfdc from "rfdc";
import addFormats from "ajv-formats";
import { AjvCompileOptions, schema_opts } from "./ajv_options.js";
import { transformCJS } from "../utils/code/cjs_to_esm.js";

const clone = rfdc();

const debug = createDebug("files");

export interface AjvFilesStoreOptions {
    ajvInstances: Record<string, Ajv>;
    resolveModule?(module: Record<string, unknown>, file: string): Record<string, any>;
    resolveSchema?(schema: any, source: string): any;
}

export interface SchemaMeta {
    options: AjvCompileOptions;
    [key: string]: any;
}

export function createAjvFileStore(opts: AjvFilesStoreOptions) {
    const {
        ajvInstances: instances,
        resolveSchema = (schema) => schema,
        resolveModule = (mod) => mod,
    } = opts;

    const id_key = "$id";

    const meta_prop = "$$meta";

    const files = new Map<string, Map<string, any>>();

    const instance_entries = Object.entries(instances);

    const default_instance = instance_entries[0]?.[1];
    if (!default_instance) {
        throw new Error("expecting at least on ajv instance");
    }

    function getFileSchemas(file: string) {
        return files.get(removeSchemaFileExt(file));
    }

    // multiple instances so we can have different options for the server and the client
    // all schemas are stored in each instance
    function forEachInstance(cb: (value: [string, Ajv]) => void) {
        return instance_entries.forEach(cb);
    }

    function addSchema(schema: any, ref: string) {
        return forEachInstance(([, ajv]) => ajv.addSchema(schema, ref));
    }

    function removeSchema(ref: string) {
        return forEachInstance(([, ajv]) => ajv.removeSchema(ref));
    }

    function removeFileSchemas(file: string) {
        const file_schemas = getFileSchemas(file);

        if (!file_schemas) {
            return;
        }

        for (const [ref, schema] of file_schemas.entries()) {
            if (schema[id_key] != undefined) {
                removeSchema(schema[id_key]);
            }

            removeSchema(resolveSchemaRef(file, ref));
        }

        files.delete(removeSchemaFileExt(file));
    }

    async function loadFileSchemas(file: string, module: Record<string, unknown>) {
        const schemas = await resolveModule(module, file);
        const schemas_entries = Object.entries(schemas);

        if (schemas_entries.length === 0) {
            console.log(yellow(`Warning: ${file}: does not export any schema`));
        }

        removeFileSchemas(file);

        const file_schemas = new Map<string, any>();

        files.set(removeSchemaFileExt(file), file_schemas);

        for (const [export_name, raw_schema] of schemas_entries) {
            const schema = await resolveSchema(clone(raw_schema), file);

            if (schema == undefined) {
                continue;
            }

            if (typeof schema != "object") {
                // idk why ajv allows boolean as schemas?
                throw new TypeError("Schema must be object");
            }

            const ref = resolveSchemaRef(file, export_name);

            const schema_clone = { ...schema };

            // see ajv_options.ts for context
            const meta = schema_clone[meta_prop];
            if (meta?.options) {
                schema_opts.set(schema_clone, meta.options);
            }

            delete schema_clone[meta_prop];

            addSchema(schema_clone, ref);
            file_schemas.set(export_name, schema);
        }

        return [...file_schemas.entries()];
    }

    /** Generate validation code for all schemas in a file */
    function getSchemaFileCode(instance: string, file: string) {
        const ajv = ensureInstance(instance);

        const file_schemas = getFileSchemas(file);
        if (!file_schemas) {
            return;
        }

        const refs = Array.from(file_schemas, ([ref]) => {
            return [ref, resolveSchemaRef(file, ref)];
        });

        const code = generateAjvStandaloneCode(ajv, Object.fromEntries(refs))
            .replace("export const default =", "export default")
            .replace('"use strict";', "");

        return transformCJS(code);
    }

    /** Generate a schema validation code */
    function getSchemaCode(ref: string, instance: string) {
        const ajv = ensureInstance(instance);
        const schema = ajv.getSchema(ref);
        if (!schema) {
            throw new Error(`schema with key or $id: ${ref} is not found`);
        }

        const code = generateAjvStandaloneCode(ajv, schema);
        return transformCJS(code);
    }

    /** Generate code represent all defined schema in a file */
    function getFileJsonSchemas(
        file: string,
        /** server=true, return a json schema with extra information($$meta)
         *
         * server=false, return the json schema as stored in ajv and without $$meta prop
         * to avoid leaking schema information to client
         */
        server = false
    ) {
        const file_schemas = getFileSchemas(file);
        if (!file_schemas) {
            return;
        }

        const schemas = [...file_schemas.entries()].map(([ref, schema]) => {
            if (server) {
                return [ref, schema];
            }
            const full_ref = resolveSchemaRef(file, ref);
            const ajv_schema = default_instance.getSchema(full_ref)?.schema;
            return [ref, ajv_schema];
        });

        return Object.fromEntries(schemas);
    }

    function ensureInstance(instance: string) {
        const ajv = instances[instance];
        if (!ajv) {
            throw Error(`Could not find ajv instance "${instance}"`);
        }
        return ajv;
    }

    function getSchemasWithId() {
        return Object.values(default_instance.schemas).filter((schema) => {
            return !schema?.meta && (schema?.schema as any)?.$id;
        });
    }

    return {
        instances,
        files,
        getFileJsonSchemas,
        getSchemaFileCode,
        getSchemaCode,
        getFileSchemas,
        removeFileSchemas,
        loadFileSchemas,
        getSchemasWithId,
    };
}

export function initInstances(instances: Record<string, Ajv>) {
    for (const instance of Object.values(instances)) {
        addFormats(instance, [
            "date-time",
            "time",
            "date",
            "email",
            "hostname",
            "ipv4",
            "ipv6",
            "uri",
            "uri-reference",
            "uuid",
            "uri-template",
            "json-pointer",
            "relative-json-pointer",
            "regex",
            /** not actual format but a hint for the UI */
            "password",
        ]);
    }
}

export function resolveSchemaRef(file: string, ref: string) {
    return `file://${removeSchemaFileExt(file)}#${ref}`;
}

export const enforcedAjvOptions: AjvOptions = {
    code: {
        esm: true,
        source: true,
        optimize: 2,
        lines: true, // REMOVEME for dev mode only
    },

    logger: {
        // currently error is not used because strict=true
        error: (...args: any[]) => console.error(red("ajv error: "), ...args),
        warn: (...args: any[]) => console.log(yellow("ajv warning: "), ...args),
        log: console.log,
    },
};

/** server options optmized for speed */
export const ajvOptionsServer: AjvOptions = {
    allErrors: false,
    removeAdditional: true,
    useDefaults: true,
    coerceTypes: true,
};

/** optmized for size */
export const ajvOptionsClient: AjvOptions = {
    inlineRefs: false,
    allErrors: true,
    useDefaults: false,
    coerceTypes: false,
    loopRequired: 4,
    loopEnum: 4,
};
