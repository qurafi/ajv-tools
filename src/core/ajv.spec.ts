import Ajv from "ajv";
import { unlinkSync, writeFileSync } from "node:fs";
import { expect, it, suite } from "vitest";
import { createAjvFileStore, enforcedAjvOptions } from "./ajv";
import path from "node:path";

suite("createAjvFileStore", () => {
    const modules = {
        "s/a.schema.js": {
            default: {},
            a: {
                $id: "a",
                type: "object",
                properties: {
                    foo: { type: "string" },
                },
                required: ["foo"],
            },
            to_be_removed: {
                $id: "to_be_removed",
            },
        },
        "s/b.schema.js": {
            default: {},
            b: {}, // no id
        },
        "s/c.schema.js": {
            default: {
                $id: "c",
                type: "object",
                properties: {
                    a: { $ref: "a" },
                },
                required: ["a"],
            },
        },
    };

    const ajv = new Ajv(enforcedAjvOptions);
    const resolved_symbol = Symbol("resolved");
    const resolved_files: string[] = [];

    const store = createAjvFileStore({
        ajvInstances: { default: ajv },
        resolveModule(module, file) {
            resolved_files.push(file);
            return module;
        },
        resolveSchema(schema) {
            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            // @ts-ignore
            schema[resolved_symbol] = true;
            return schema;
        },
    });

    Object.entries(modules).map(([file, module]) => {
        return store.loadFileSchemas(file, module);
    });

    it("should load all schemas", () => {
        expect(Object.entries(ajv.schemas).slice(1)).toMatchSnapshot();
    });

    it.todo("should remove old schemas when loading the same file");

    it.todo("should throw when adding multiple ids from different files");

    it("should use resolveModule", () => {
        expect(resolved_files).toStrictEqual(Object.keys(modules));
    });

    it("should loadd all schemas", () => {
        const schemas = store.getFileSchemas("s/a.schema.js");
        console.log({ schemas });
    });

    it("should validate correctly", () => {
        const schema_a = ajv.getSchema("a");
        expect(schema_a).toBeDefined();

        if (!schema_a) return;

        const valid_0 = schema_a({ foo: "correct" });

        expect(valid_0).toBe(true);
        expect(schema_a?.errors).toBe(null);

        const valid_1 = schema_a({});
        expect(valid_1).toBe(false);
        expect(schema_a?.errors).toMatchSnapshot();
    });

    it("should validate cross-file reference", () => {
        const schema_a = ajv.getSchema("c");
        expect(schema_a).toBeDefined();

        if (!schema_a) return;

        const valid_0 = schema_a({ a: { foo: "correct" } });
        expect(valid_0).toBe(true);
        expect(schema_a?.errors).toBe(null);

        const valid_1 = schema_a({ a: { foo: 1 } });
        expect(valid_1).toBe(false);
        expect(schema_a?.errors).toMatchSnapshot();

        const valid_2 = schema_a({ a: {} });
        expect(valid_2).toBe(false);
        expect(schema_a?.errors).toMatchSnapshot();
    });

    it("should resolve schema with resolveSchema", () => {
        const schema_a = ajv.getSchema("a");
        expect(schema_a).toBeDefined();

        const { schema } = schema_a!;

        expect(typeof schema == "object" && resolved_symbol in schema).toBe(true);
    });

    it("getSchemaFileCode should return the code and validate correctly", async () => {
        // console.log(store.files);
        console.log(Object.keys(ajv.schemas));
        // console.log(ajv.getSchema("file://s/a.schema/default"));
        const code = store.getSchemaFileCode("default", "s/a.schema");
        expect(code).toBeDefined();

        const tmp_path = path.resolve(__dirname, `./schema.tmp.${Date.now()}.js`);

        writeFileSync(tmp_path, code!);

        const module = await import(tmp_path);

        unlinkSync(tmp_path);

        // yeah duplicate code from the above, we need some refactoring later
        // TODO
        const valid_0 = module.a({ foo: "correct" });

        expect(valid_0).toBe(true);
        expect(module.a?.errors).toBe(null);

        const valid_1 = module.a({});
        expect(valid_1).toBe(false);
        expect(module.a?.errors).toMatchSnapshot();

        expect(Object.keys(module)).toStrictEqual(Object.keys(modules["s/a.schema.js"]));
    });

    it.todo("should be fine with cross-file reference with any order");
    it.todo("same file schema reference tests without id using file:// ref");
    it.todo("cross-file schema reference tests without id using file:// ref");
});
