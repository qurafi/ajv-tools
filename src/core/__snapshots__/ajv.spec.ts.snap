// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createAjvFileStore > getSchemaFileCode should return the code and validate correctly 1`] = `
[
  {
    "instancePath": "",
    "keyword": "required",
    "message": "must have required property 'foo'",
    "params": {
      "missingProperty": "foo",
    },
    "schemaPath": "#/required",
  },
]
`;

exports[`createAjvFileStore > should load all schemas 1`] = `
[
  [
    "http://json-schema.org/draft-07/schema",
    SchemaEnv {
      "$async": undefined,
      "baseId": "http://json-schema.org/draft-07/schema",
      "dynamicAnchors": {},
      "localRefs": {},
      "meta": true,
      "refs": {
        "http://json-schema.org/draft-07/schema#/definitions/nonNegativeInteger": {
          "minimum": 0,
          "type": "integer",
        },
        "http://json-schema.org/draft-07/schema#/definitions/nonNegativeIntegerDefault0": SchemaEnv {
          "$async": undefined,
          "baseId": "http://json-schema.org/draft-07/schema",
          "dynamicAnchors": {},
          "localRefs": undefined,
          "meta": undefined,
          "refs": {},
          "root": [Circular],
          "schema": {
            "allOf": [
              {
                "$ref": "#/definitions/nonNegativeInteger",
              },
              {
                "default": 0,
              },
            ],
          },
          "schemaId": "$id",
          "schemaPath": undefined,
          "validate": [Function],
          "validateName": ValueScopeName {
            "prefix": "validate",
            "scopePath": _Code {
              "_items": [
                ".",
                Name {
                  "str": "validate",
                },
                "[",
                0,
                "]",
              ],
            },
            "str": "validate1",
            "value": {
              "ref": [Function],
            },
          },
        },
        "http://json-schema.org/draft-07/schema#/definitions/schemaArray": SchemaEnv {
          "$async": undefined,
          "baseId": "http://json-schema.org/draft-07/schema",
          "dynamicAnchors": {},
          "localRefs": undefined,
          "meta": undefined,
          "refs": {},
          "root": [Circular],
          "schema": {
            "items": {
              "$ref": "#",
            },
            "minItems": 1,
            "type": "array",
          },
          "schemaId": "$id",
          "schemaPath": undefined,
          "validate": [Function],
          "validateName": ValueScopeName {
            "prefix": "validate",
            "scopePath": _Code {
              "_items": [
                ".",
                Name {
                  "str": "validate",
                },
                "[",
                1,
                "]",
              ],
            },
            "str": "validate3",
            "value": {
              "ref": [Function],
            },
          },
        },
        "http://json-schema.org/draft-07/schema#/definitions/simpleTypes": {
          "enum": [
            "array",
            "boolean",
            "integer",
            "null",
            "number",
            "object",
            "string",
          ],
        },
        "http://json-schema.org/draft-07/schema#/definitions/stringArray": {
          "default": [],
          "items": {
            "type": "string",
          },
          "type": "array",
          "uniqueItems": true,
        },
        "https://raw.githubusercontent.com/ajv-validator/ajv/master/lib/refs/data.json": {
          "$id": "https://raw.githubusercontent.com/ajv-validator/ajv/master/lib/refs/data.json#",
          "additionalProperties": false,
          "description": "Meta-schema for $data reference (JSON AnySchema extension proposal)",
          "properties": {
            "$data": {
              "anyOf": [
                {
                  "format": "relative-json-pointer",
                },
                {
                  "format": "json-pointer",
                },
              ],
              "type": "string",
            },
          },
          "required": [
            "$data",
          ],
          "type": "object",
        },
      },
      "root": [Circular],
      "schema": {
        "$id": "http://json-schema.org/draft-07/schema#",
        "$schema": "http://json-schema.org/draft-07/schema#",
        "default": true,
        "definitions": {
          "nonNegativeInteger": {
            "minimum": 0,
            "type": "integer",
          },
          "nonNegativeIntegerDefault0": {
            "allOf": [
              {
                "$ref": "#/definitions/nonNegativeInteger",
              },
              {
                "default": 0,
              },
            ],
          },
          "schemaArray": {
            "items": {
              "$ref": "#",
            },
            "minItems": 1,
            "type": "array",
          },
          "simpleTypes": {
            "enum": [
              "array",
              "boolean",
              "integer",
              "null",
              "number",
              "object",
              "string",
            ],
          },
          "stringArray": {
            "default": [],
            "items": {
              "type": "string",
            },
            "type": "array",
            "uniqueItems": true,
          },
        },
        "properties": {
          "$comment": {
            "type": "string",
          },
          "$id": {
            "format": "uri-reference",
            "type": "string",
          },
          "$ref": {
            "format": "uri-reference",
            "type": "string",
          },
          "$schema": {
            "format": "uri",
            "type": "string",
          },
          "additionalItems": {
            "$ref": "#",
          },
          "additionalProperties": {
            "$ref": "#",
          },
          "allOf": {
            "$ref": "#/definitions/schemaArray",
          },
          "anyOf": {
            "$ref": "#/definitions/schemaArray",
          },
          "const": {
            "anyOf": [
              true,
              {
                "$ref": "https://raw.githubusercontent.com/ajv-validator/ajv/master/lib/refs/data.json#",
              },
            ],
          },
          "contains": {
            "$ref": "#",
          },
          "contentEncoding": {
            "type": "string",
          },
          "contentMediaType": {
            "type": "string",
          },
          "default": true,
          "definitions": {
            "additionalProperties": {
              "$ref": "#",
            },
            "default": {},
            "type": "object",
          },
          "dependencies": {
            "additionalProperties": {
              "anyOf": [
                {
                  "$ref": "#",
                },
                {
                  "$ref": "#/definitions/stringArray",
                },
              ],
            },
            "type": "object",
          },
          "description": {
            "type": "string",
          },
          "else": {
            "$ref": "#",
          },
          "enum": {
            "anyOf": [
              {
                "items": true,
                "minItems": 1,
                "type": "array",
                "uniqueItems": true,
              },
              {
                "$ref": "https://raw.githubusercontent.com/ajv-validator/ajv/master/lib/refs/data.json#",
              },
            ],
          },
          "examples": {
            "items": true,
            "type": "array",
          },
          "exclusiveMaximum": {
            "anyOf": [
              {
                "type": "number",
              },
              {
                "$ref": "https://raw.githubusercontent.com/ajv-validator/ajv/master/lib/refs/data.json#",
              },
            ],
          },
          "exclusiveMinimum": {
            "anyOf": [
              {
                "type": "number",
              },
              {
                "$ref": "https://raw.githubusercontent.com/ajv-validator/ajv/master/lib/refs/data.json#",
              },
            ],
          },
          "format": {
            "anyOf": [
              {
                "type": "string",
              },
              {
                "$ref": "https://raw.githubusercontent.com/ajv-validator/ajv/master/lib/refs/data.json#",
              },
            ],
          },
          "if": {
            "$ref": "#",
          },
          "items": {
            "anyOf": [
              {
                "$ref": "#",
              },
              {
                "$ref": "#/definitions/schemaArray",
              },
            ],
            "default": true,
          },
          "maxItems": {
            "anyOf": [
              {
                "$ref": "#/definitions/nonNegativeInteger",
              },
              {
                "$ref": "https://raw.githubusercontent.com/ajv-validator/ajv/master/lib/refs/data.json#",
              },
            ],
          },
          "maxLength": {
            "anyOf": [
              {
                "$ref": "#/definitions/nonNegativeInteger",
              },
              {
                "$ref": "https://raw.githubusercontent.com/ajv-validator/ajv/master/lib/refs/data.json#",
              },
            ],
          },
          "maxProperties": {
            "anyOf": [
              {
                "$ref": "#/definitions/nonNegativeInteger",
              },
              {
                "$ref": "https://raw.githubusercontent.com/ajv-validator/ajv/master/lib/refs/data.json#",
              },
            ],
          },
          "maximum": {
            "anyOf": [
              {
                "type": "number",
              },
              {
                "$ref": "https://raw.githubusercontent.com/ajv-validator/ajv/master/lib/refs/data.json#",
              },
            ],
          },
          "minItems": {
            "anyOf": [
              {
                "$ref": "#/definitions/nonNegativeIntegerDefault0",
              },
              {
                "$ref": "https://raw.githubusercontent.com/ajv-validator/ajv/master/lib/refs/data.json#",
              },
            ],
          },
          "minLength": {
            "anyOf": [
              {
                "$ref": "#/definitions/nonNegativeIntegerDefault0",
              },
              {
                "$ref": "https://raw.githubusercontent.com/ajv-validator/ajv/master/lib/refs/data.json#",
              },
            ],
          },
          "minProperties": {
            "anyOf": [
              {
                "$ref": "#/definitions/nonNegativeIntegerDefault0",
              },
              {
                "$ref": "https://raw.githubusercontent.com/ajv-validator/ajv/master/lib/refs/data.json#",
              },
            ],
          },
          "minimum": {
            "anyOf": [
              {
                "type": "number",
              },
              {
                "$ref": "https://raw.githubusercontent.com/ajv-validator/ajv/master/lib/refs/data.json#",
              },
            ],
          },
          "multipleOf": {
            "anyOf": [
              {
                "exclusiveMinimum": 0,
                "type": "number",
              },
              {
                "$ref": "https://raw.githubusercontent.com/ajv-validator/ajv/master/lib/refs/data.json#",
              },
            ],
          },
          "not": {
            "$ref": "#",
          },
          "oneOf": {
            "$ref": "#/definitions/schemaArray",
          },
          "pattern": {
            "anyOf": [
              {
                "format": "regex",
                "type": "string",
              },
              {
                "$ref": "https://raw.githubusercontent.com/ajv-validator/ajv/master/lib/refs/data.json#",
              },
            ],
          },
          "patternProperties": {
            "additionalProperties": {
              "$ref": "#",
            },
            "default": {},
            "propertyNames": {
              "format": "regex",
            },
            "type": "object",
          },
          "properties": {
            "additionalProperties": {
              "$ref": "#",
            },
            "default": {},
            "type": "object",
          },
          "propertyNames": {
            "$ref": "#",
          },
          "readOnly": {
            "default": false,
            "type": "boolean",
          },
          "required": {
            "anyOf": [
              {
                "$ref": "#/definitions/stringArray",
              },
              {
                "$ref": "https://raw.githubusercontent.com/ajv-validator/ajv/master/lib/refs/data.json#",
              },
            ],
          },
          "then": {
            "$ref": "#",
          },
          "title": {
            "type": "string",
          },
          "type": {
            "anyOf": [
              {
                "$ref": "#/definitions/simpleTypes",
              },
              {
                "items": {
                  "$ref": "#/definitions/simpleTypes",
                },
                "minItems": 1,
                "type": "array",
                "uniqueItems": true,
              },
            ],
          },
          "uniqueItems": {
            "anyOf": [
              {
                "default": false,
                "type": "boolean",
              },
              {
                "$ref": "https://raw.githubusercontent.com/ajv-validator/ajv/master/lib/refs/data.json#",
              },
            ],
          },
        },
        "title": "Core schema meta-schema",
        "type": [
          "object",
          "boolean",
        ],
      },
      "schemaId": "$id",
      "schemaPath": undefined,
      "validate": [Function],
      "validateName": ValueScopeName {
        "prefix": "validate",
        "scopePath": _Code {
          "_items": [
            ".",
            Name {
              "str": "validate",
            },
            "[",
            2,
            "]",
          ],
        },
        "str": "validate0",
        "value": {
          "ref": [Function],
        },
      },
    },
  ],
  [
    "file://s/a.schema/default.json",
    SchemaEnv {
      "$async": undefined,
      "baseId": "file://s/a.schema/default.json",
      "dynamicAnchors": {},
      "localRefs": {},
      "meta": undefined,
      "refs": {},
      "root": [Circular],
      "schema": {
        Symbol(resolved): true,
      },
      "schemaId": "$id",
      "schemaPath": undefined,
    },
  ],
  [
    "file://s/b.schema/default.json",
    SchemaEnv {
      "$async": undefined,
      "baseId": "file://s/b.schema/default.json",
      "dynamicAnchors": {},
      "localRefs": {},
      "meta": undefined,
      "refs": {},
      "root": [Circular],
      "schema": {
        Symbol(resolved): true,
      },
      "schemaId": "$id",
      "schemaPath": undefined,
    },
  ],
  [
    "file://s/c.schema/default.json",
    SchemaEnv {
      "$async": undefined,
      "baseId": "c",
      "dynamicAnchors": {},
      "localRefs": {},
      "meta": undefined,
      "refs": {},
      "root": [Circular],
      "schema": {
        "$id": "c",
        "properties": {
          "a": {
            "$ref": "a",
          },
        },
        "required": [
          "a",
        ],
        "type": "object",
        Symbol(resolved): true,
      },
      "schemaId": "$id",
      "schemaPath": undefined,
    },
  ],
  [
    "file://s/a.schema/a.json",
    SchemaEnv {
      "$async": undefined,
      "baseId": "a",
      "dynamicAnchors": {},
      "localRefs": {},
      "meta": undefined,
      "refs": {},
      "root": [Circular],
      "schema": {
        "$id": "a",
        "properties": {
          "foo": {
            "type": "string",
          },
        },
        "required": [
          "foo",
        ],
        "type": "object",
        Symbol(resolved): true,
      },
      "schemaId": "$id",
      "schemaPath": undefined,
    },
  ],
  [
    "file://s/b.schema/b.json",
    SchemaEnv {
      "$async": undefined,
      "baseId": "file://s/b.schema/b.json",
      "dynamicAnchors": {},
      "localRefs": {},
      "meta": undefined,
      "refs": {},
      "root": [Circular],
      "schema": {
        Symbol(resolved): true,
      },
      "schemaId": "$id",
      "schemaPath": undefined,
    },
  ],
  [
    "file://s/a.schema/to_be_removed.json",
    SchemaEnv {
      "$async": undefined,
      "baseId": "to_be_removed",
      "dynamicAnchors": {},
      "localRefs": {},
      "meta": undefined,
      "refs": {},
      "root": [Circular],
      "schema": {
        "$id": "to_be_removed",
        Symbol(resolved): true,
      },
      "schemaId": "$id",
      "schemaPath": undefined,
    },
  ],
]
`;

exports[`createAjvFileStore > should validate correctly 1`] = `
[
  {
    "instancePath": "",
    "keyword": "required",
    "message": "must have required property 'foo'",
    "params": {
      "missingProperty": "foo",
    },
    "schemaPath": "#/required",
  },
]
`;

exports[`createAjvFileStore > should validate cross-file reference 1`] = `
[
  {
    "instancePath": "/a/foo",
    "keyword": "type",
    "message": "must be string",
    "params": {
      "type": "string",
    },
    "schemaPath": "a/properties/foo/type",
  },
]
`;

exports[`createAjvFileStore > should validate cross-file reference 2`] = `
[
  {
    "instancePath": "/a",
    "keyword": "required",
    "message": "must have required property 'foo'",
    "params": {
      "missingProperty": "foo",
    },
    "schemaPath": "a/required",
  },
]
`;
